lib/utils/remote-build.ts

```diff
@@ -216,72 +213,120 @@ async function cancelBuildIfNecessary(build: RemoteBuild): Promise<void> {
+		tarSpinner.stop();
+	}
+}
 
+/**
+ * Initiate a POST HTTP request to the remote builder and add some event
+ * listeners.
+ *
+ * ¡! Note: this function must be synchronous because of a bug in the `request`
+ *    library that requires the following two steps to take place in the same
+ *    iteration of Node's event loop: (1) adding a listener for the 'response'
+ *    event and (2) calling request.pipe():
+ *    https://github.com/request/request/issues/887
+ */
+function createRemoteBuildRequest(
+	build: RemoteBuild,
+	tarStream: Stream.Readable,
+	builderUrl: string,
+	onError: (error: Error) => void,
+): request.Request {
+	const zlib = require('zlib');
 	if (DEBUG_MODE) {
-		console.log(`[debug] Connecting to builder at ${url}`);
+		console.log(`[debug] Connecting to builder at ${builderUrl}`);
 	}
-	const post = request.post({
-		url,
-		auth: {
-			bearer: build.auth,
-		},
-		headers: {
-			'Content-Encoding': 'gzip',
-		},
-		body: tarStream.pipe(
-			zlib.createGzip({
-				level: 6,
-			}),
-		),
-	});
+	return request
+		.post({
+			url: builderUrl,
+			auth: { bearer: build.auth },
+			headers: { 'Content-Encoding': 'gzip' },
+			body: tarStream.pipe(zlib.createGzip({ level: 6 })),
+		})
+		.on('error', onError)
+		.once('response', (response: request.RequestResponse) => {
+			if (
+				response.statusCode &&
+				/^[23]\d\d$/.test(response.statusCode.toString())
+			) {
+				if (DEBUG_MODE) {
+					console.log(
+						`[debug] received HTTP ${response.statusCode} ${
+							response.statusMessage
+						}`,
+					);
+				}
+			} else {
+				let msgArr = [
+					'Remote builder responded with HTTP error:',
+					`${response.statusCode} ${response.statusMessage}`,
+				];
+				if (response.body) {
+					msgArr.push(response.body);
+				}
+				onError(new Error(msgArr.join('\n')));
+			}
+		});
+}
 
+async function getRemoteBuildStream(
+	build: RemoteBuild,
+): Promise<NodeJS.ReadWriteStream> {
+	const tarStream = await getTarStream(build);
+	const visuals = await import('resin-cli-visuals');
 	const uploadSpinner = new visuals.Spinner(
 		'Uploading source package to balena cloud',
 	);
-	uploadSpinner.start();
+	const exitOnError = (error: Error): never => {
+		uploadSpinner.stop();
+		return exitWithExpectedError(error);
+	};
 
-	const parseStream = post.pipe(JSONStream.parse('*'));
-	parseStream.on('data', () => uploadSpinner.stop());
-	return parseStream as Stream.Duplex;
+	try {
+		uploadSpinner.start();
+		const builderUrl = await getBuilderEndpoint(
+			build.baseUrl,
+			build.owner,
+			build.app,
+			build.opts,
+		);
+		const buildRequest = createRemoteBuildRequest(
+			build,
+			tarStream,
+			builderUrl,
+			exitOnError,
+		);
+		return buildRequest
+			.pipe(JSONStream.parse('*'))
+			.once('data', () => uploadSpinner.stop())
+			.once('end', () => uploadSpinner.stop())
+			.once('close', () => uploadSpinner.stop())
+			.once('error', () => uploadSpinner.stop());
+	} catch (error) {
+		return exitOnError(error);
```

Comment:
Previously, in some cases if the builder request returned a HTTP error code, the CLI would not print any error message and would get stuck showing the spinner forever...